### ISSUE TITLE
agda-mode: generate with-clauses from hole

### ISSUE THREAD

**Issue by m0rphism** [agda/agda#2768]
I propose adding a feature which given

    foo : ℕ → ℕ → ℕ
    foo x y = {! x ≟ y !}

generates

    foo : ℕ → ℕ → ℕ
    foo x y with x ≟ y
    ... | c = {!!}

and ideally moves the cursor to the new hole.

There is an implementation of that feature in [this](https://github.com/wjzz/Agda-mode-improvements) repository. I've tested the lisp file and it still seems to work, even though the last commit was 6 years ago and the user hasn't had any github activity since 2014.

**Comment by nad** [comment]
Duplicate of #560.

**Comment by andreasabel** [comment]
Thanks, @m0rphism, for the pointer.  Would you be able to turn this into a pull request?  On the Agda side, the new functionality would have to be added to `Agda.Interaction.InteractionTop`.

**Comment by m0rphism** [comment]
I could give it a try, but I have no experience with the Agda codebase yet and also not much experience with emacs lisp, so I may need a bit of assistance to get things going.

I had a closer look at the existing implemention I've mentioned, which is an elisp-only solution.

The good thing is, that it can also create with-clauses matching on multiple expressions, e.g.

    foo : ℕ → ℕ → ℕ
    foo x y = {! x ≟ y | y ≟ x !}

expands to

    foo : ℕ → ℕ → ℕ
    foo x y with x ≟ y | y ≟ x
    ... | cond0 | cond1 = {!!}

and correctly places the cursor in the new hole.

The bad thing is, that it breaks in two scenarios
1.  It requires the whole `lhs = rhs-with-hole` to be on the same line, otherwise indentation and fresh variable selection goes wrong, e.g.

        foo : ℕ → ℕ → ℕ
        foo cond1
            cond0 = {! cond0 ≟ cond1 !}

    expands to

        foo : ℕ → ℕ → ℕ
        foo cond1
            cond0 with cond0 ≟ cond1
            ... | cond1 = {!!}

2.  Sometimes fresh variable selection goes wrong even if the condition from item 1 is met, e.g.

        foo cond1 cond0 = {! (cond0 ≟ cond1)!}

    expands to

        foo cond1 cond0 with (cond0 ≟ cond1)
        ... | cond1 = {!!}

While it should be possible to fix those issues in elisp, it would probably be a lot more robust to make use of the Agda compiler instead and extend `Agda.Interaction.InteractionTop`. 

I'll try to find some time to get a bit more familiar with the agda-mode / agda-compiler interaction, maybe I can replace at least parts of the elisp-only solution by interaction with the Agda compiler.

**Comment by gallais** [comment]
I don't think this needs to be perfect to be included. A lot of the current agda-mode features already break when things are not on the same line. Using `C-c C-c` in the following hole for instance leads to an invalid file:

```agda
open import Agda.Builtin.Nat

pred : Nat → Nat
pred
  n = {! n !}
```

However, @wjzz's work is not explicitly permissively licensed so I guess that integrating it as is is not an option at least until we have acquired his consent.

**Comment by wjzz** [comment]
Feel free to use the code if you find it useful :)

22 wrz 2017 17:32 "G. Allais" <notifications@github.com> napisał(a):

**Comment by UlfNorell** [comment]
Should also handle `rewrite` (from #2813).

The suggestion there was to not have a new command, but change `give` to recognise `rewrite e` and `with e` and do the right thing.

**Comment by gallais** [comment]
The parser already has a `HoleContent` target:

https://github.com/agda/agda/blob/0a284d1070563150605080c55a0ce95791290e20/src/full/Agda/Syntax/Parser/Parser.y#L1116-L1123

But we do not seem to make any use of it. Seems like we could generalise
`C-c C-c` to use this parser instead of merely looking at the content of
the hole as it if were a bunch of variables.
